src/Albany_DataTypes.hpp

```diff
@@ -59,36 +59,48 @@ typedef Sacado::Fad::DFad<RealType> TanFadType;
 #include "MatrixMarket_Tpetra.hpp"
 #include "Tpetra_RowMatrixTransposer.hpp"
 
-
 //Kokkos includes
 #include "Phalanx_KokkosDeviceTypes.hpp"
 
 //Tpetra typedefs
-typedef RealType                                    ST;
-#ifdef ALBANY_64BIT_INT
-typedef long long int                               GO;
+typedef RealType                                             ST;
+typedef std::int64_t                                         GO;
+typedef std::int32_t                                         LO;
+typedef Kokkos::Compat::KokkosDeviceWrapperNode<PHX::Device> KokkosNode;
+
+typedef int Tpetra_LO;
+#if defined( HAVE_TPETRA_INST_INT_LONG_LONG )
+typedef long long Tpetra_GO;
+#elif defined( HAVE_TPETRA_INST_INT_LONG )
+static_assert(sizeof(long) == sizeof(GO),
+    "Tpetra's biggest enabled GlobalOrdinal is long but thats not 64 bit");
+typedef long Tpetra_GO;
```

Comment:
Why bothering with two types names if we enforce them to have the same size? Can't we just have `GO`, so that people don't get confused? If we allowed Albany to be built with 32 bits global ordinals, then it would make sense (but I would check that `sizeof(Tpetra_GO)>=sizeof(GO)` then).

Reply:
Because `Tpetra_GO` is used as a template parameter to defined `Tpetra_Vector`, it has to be _exactly_ the type that Tpetra supports, not just one that is compatible.

Reply:
Oh, like `GO=long` and `Tpetra_GO=unsigned long`? But if we allow this, aren't we running into a lot of warning when, e.g., later on we compare a `GO` with a `Tpetra_GO`? Or, even worse, if we get an implicit conversion from one to the other when it is not a positive `GO`? I gave a quick look and we only convert a `GO` into a `Tpetra_GO`, which is fine if the `GO` is non-negative (and maybe even -1 is fine cause it may be caught by Tpetra as an Invalid ordinal). But aren't we walking on thin ice, with the risk of a bug sneaking in at any time? Can't we just ask for the same type, not just same size at this point? I'm really not sure, maybe I'm just overreacting.

Reply:
See @mhoemmen 's comments in #213 where he advocates that we not require exactly the same type. Yes, we should use appropriate conversions, and safe ones if possible. Note that ordinals are not actually supposed to ever be negative, there are other reasons for using signed integers even if most of the time stored values are positive. Also, this is is something we'll have to do for DTK anyways. We need to get into the habit of knowing which type each package uses and converting as needed. Its actually more painful (in fact impossible, e.g. DTK) to constantly reconfigure packages until they all have exactly the same types.

Reply:
I'm actually expecting that our most common conversion will be `long` <-> `long long`, which are the same for all practical purposes on Linux but technically different per the compiler. I think `int64_t` is `long`, and Tpetra usually goes with `long long`.

Reply:
I've seen cases where `int64_t` is `long`, and other cases where it is `long long`.  (Super annoying!  I really wish `int64_t` were a proper type instead of an alias (typedef).)

Reply:
Albany could consider using a "tagged integer" type for `GO`.  This would prevent arithmetic that implicitly mixes global and local indices, for example.

Reply:
> Albany could consider using a "tagged integer" type for `GO`.

Interesting idea... can you point to an implementation / example / description of these things so we can consider it?

Reply:
@ibaned Sierra's Aria product uses this extensively.  We've wanted this for years in Tpetra, but it would break backwards compatibility.  If an app decouples its global index type from Tpetra, it's free to add this feature.

Reply:
some quick googling turned [this](http://www.ilikebigbits.com/blog/2014/5/6/type-safe-identifiers-in-c) up... seems pretty compelling... I can think of possible unwanted side effects, but it may be worth a try.

Reply:
@ibaned Yup!  I don't mean to tell Albany that they should do this, just that if Albany decouples its global index type from Tpetra's, then Albany has the opportunity to consider things like this.

Reply:
[this](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3515.pdf) would make all of this much easier...

Reply:
@bartgol yes, tagged integers seem to be a template trick to basically implement opaque typedefs, as much as possible in the current language.